### PR TITLE
Add AuthorizationMiddleware and update sandbox

### PR DESCRIPTION
**Pull request by JamesNK**
https://github.com/aspnet/Routing/issues/537

PR adds first cut of `AuthorizationMiddleware`. Largely a copy from Ryan's original implementation, which is an adaption of `AuthorizationFilter`. Highlevel changes:
* I added a check for `IAllowAnonymous` metadata so it is more inline with MVC's `AuthorizationFilter`
* Convention builder for adding `IAuthorizeData` as metadata to endpoints. Example: https://github.com/aspnet/Routing/compare/master...jamesnk/authz-middleware?expand=1#diff-c65f61210298d7cca90708f6e5f60b46R55

Run the routing sandbox sample to try it out.

@rynowak where is the best long term home for the middleware? BasicMiddleware? An existing authorization package? New authorization package?

@blowdart It is very similar to MVC's `AuthorizationFilter`. Is this a good template for doing authz in middleware?

Will add unit tests once we're happy with the overall approach and it is moved to its final home.

### CODE REVIEW

**Comment by blowdart** [comment]
I don't get what you're trying to do here, but generally everything would go though the IAuthoizationService rather than calling evaluate on the policy directory, as that way people can swap the entire thing out for something like PolicyServer

**Comment by rynowak** [comment]
> @rynowak where is the best long term home for the middleware? BasicMiddleware? An existing authorization package? New authorization package? @blowdart It is very similar to MVC's AuthorizationFilter. Is this a good template for doing authz in middleware?

I think it's probably worth a short meeting to discuss these kinds of details.

File: samples/Microsoft.AspNetCore.Authorization.Middleware/AuthorizationEndpointConventionBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    public static class AuthorizationEndpointConventionBuilderExtensions
+    {
+        // REVIEW: What other overloads should exist?
+        // e.g. params string[] policies, or params string[] roles
+        public static IEndpointConventionBuilder RequireAuthorization(this IEndpointConventionBuilder builder, params IAuthorizeData[] authorizeData)
```
[the remark this answers is not in the document]
Comment (rynowak) [reply samples/Microsoft.AspNetCore.Authorization.Middleware/AuthorizationEndpointConventionBuilderExtensions.cs]: It all ships in a single box.

**Comment by rynowak** [comment]
> I don't get what you're trying to do here, but generally everything would go though the IAuthoizationService rather than calling evaluate on the policy directory, as that way people can swap the entire thing out for something like PolicyServer

@blowdart do you have a sample of this somewhere? This code was lifted pretty much verbatim from MVC so if it's wrong here then it's wrong there.

**Comment by JamesNK** [comment]
Replaced with https://github.com/aspnet/Security/pull/1894

**Pull request closed without merging**